mysql-test/mysql-test-run.pl

```diff
@@ -929,9 +929,9 @@ ($$$)
 
                 rename $log_file_name, $log_file_name.".failed";
               }
-	      delete($result->{result});
 	      $result->{retries}= $retries+1;
 	      $result->write_test($sock, 'TESTCASE');
+              push(@$completed, $result);
 	      next;
```

Comment:
okay. that kind of make sense if you're going to treat all retries as independent test runs. I think this is the correct thing to do. But then I don't understand why did you *also* added that mess with fake xml-only statuses.

Reply:
Statuses in XML copy most of the statuses in the log, there is almost complete correspondence between the log and XML.

Reply:
I guess you need to catch number of retries for each test case that is repeated, where starting number of retries is 2. So here you don't catch the first one, right? I guess you can write this as:
```
--- a/mysql-test/mysql-test-run.pl
+++ b/mysql-test/mysql-test-run.pl
@@ -888,6 +888,7 @@ sub run_test_server ($$$) {
 
                 rename $log_file_name, $log_file_name.".failed";
               }
+           push(@$completed, $result);
              delete($result->{result});
              $result->{retries}= $retries+1;
              $result->write_test($sock, 'TESTCASE');
```

Reply:
If we do not delete `delete($result->{result})`, then the status of the second successful test will be considered not successful.

Reply:
- It is about incrementing of the retries (first test will have +1 saved what should have next test)
- If the test is successful it will not go through this part of code - retries are done for test that is failed, so I guess it is ok just to apply suggested patch, when you delete the result, new result will be created when new (retry) test started.

Reply:
It works differently now. Let's assume that the number of attempts is 3. Then the following situation may occur:
* Failed
* Succeed
* Succeed

In this situation, when deleting a result, the second one will be written as Failed.

Reply:
I don't see how is possible to have single test that 2 times is successful :). Let's see:
```
	  # Retry test run after test failure
	  my $retries= $result->{retries} || 2;
	  my $test_has_failed= $result->{failures} || 0;
	  if ($test_has_failed and $retries <= $opt_retry){
```
Number of attempts is 3 (`opt_retry = 3`, default value is 1) that means it will be repeated only 2 times:
1. iteration retries = 2 and test is failed  (failure > 0) (result = MTR_RES_FAILED) , ok there is condition to repeat the test. With my suggestion (`retries = 2`) will be saved and result deleted),  increment retries (retries=3). With your patch you are saving 3 as number of retries, not ok. Run failed test again. When the test is run new `result` will be created so deleting should be optional, but I would leave it as it was.
2. iteration retries=3, 
  - result= MTR_RES_FAILED - > same as above, save retries=3 and result failed.No more repeating.
  - result= MTR_RES_PASSED ->we will not go in the if condition and test will finish.
Can you please show me the result you claim that in your generated xml-report you have 2 successful tests with retries > 2 and also on stdout?

Reply:
The logic with restarts even after successful ones is in the --retry option. The --retry option is responsible for how many restarts to make after the first failed test. The --retry-failure option is responsible for how many maximum failure attempts there will be for this test. https://github.com/MariaDB/server/blob/10.2/mysql-test/mysql-test-run.pl#L866

Reply:
@DmitriyKarpovskiy1 , @timofeyturenko  please try with following patch instead of above:
```

--- a/mysql-test/mysql-test-run.pl
+++ b/mysql-test/mysql-test-run.pl
@@ -888,6 +888,7 @@ sub run_test_server ($$$) {
 
                 rename $log_file_name, $log_file_name.".failed";
               }
+          $result->{retries}= $retries;
+           push(@$completed, $result);
              delete($result->{result});
              $result->{retries}= $retries+1;
              $result->write_test($sock, 'TESTCASE');
```
Before storing the test case you have to store valid number of retries. 
Since again new test case is executed new result will be obtained, so it is optional IMO deleting the result.

Reply:
Your code is not working correctly. Generated XML:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites disabled="0" errors="" failures="4" name="" tests="4" time="27.998">
        <testsuite disabled="" errors="" failures="" hostname="mdbci-2g4f9e2o-1616481906-node" id="0" name="galera_3nodes" package="" skipped="" tests="4" time="27.998" timestamp="2021-03-25T15:44:34Z">
                <testcase assertions="" classname="galera_3nodes" name="galera_ipv6_mariabackup" status="MTR_RES_FAILED" time="0.000" combinations="innodb" />
                <testcase assertions="" classname="galera_3nodes" name="galera_ipv6_mariabackup" status="MTR_RES_RETRY_FAILED" time="0.000" combinations="innodb" />
                <testcase assertions="" classname="galera_3nodes" name="galera_ipv6_mariabackup" status="MTR_RES_RETRY_FAILED" time="14.364" combinations="innodb" />
                <testcase assertions="" classname="galera_3nodes" name="galera_ipv6_mariabackup" status="MTR_RES_RETRY_PASSED" time="13.634" combinations="innodb" />
        </testsuite>
</testsuites>
```

But this is not the case. The statuses were:
1. Fail
2. Fail
3. Pass
4. Pass

At the same time, messages about fails are also incorrectly processed (they simply do not exist).

Reply:
@DmitriyKarpovskiy1  I was thinking on this change:
```
--- a/mysql-test/lib/mtr_report.pm
+++ b/mysql-test/lib/mtr_report.pm
@@ -500,13 +500,17 @@ sub mtr_report_stats ($$$$) {
       my $test_result;
 
       if ($test->{'retries'} > 0) {
-        $test_result = "MTR_RES_FAILED";
-        if ($test->{'failures'} > 1){
-          $test_result = "MTR_RES_RETRY_FAILED";
-        }
         if ($test->{'result'} eq "MTR_RES_PASSED"){
           $test_result = "MTR_RES_RETRY_PASSED";
         }
+        else{
+          if ($test->{'failures'} == 1){
+            $test_result = $test->{'result'}; # failure
+          }
+          else{
+            $test_result = "MTR_RES_RETRY_FAILED"; # retry failure
+          }
+        }
       } else {
         $test_result = $test->{'result'};
       }
diff --git a/mysql-test/mysql-test-run.pl b/mysql-test/mysql-test-run.pl
index 341d24360ec..6966ee96e88 100755
--- a/mysql-test/mysql-test-run.pl
+++ b/mysql-test/mysql-test-run.pl
@@ -888,9 +888,10 @@ sub run_test_server ($$$) {
 
                 rename $log_file_name, $log_file_name.".failed";
               }
+        $result->{retries}= $retries;
+        push(@$completed, $result);
              $result->{retries}= $retries+1;
              $result->write_test($sock, 'TESTCASE');
-             push(@$completed, $result);
              next;
            }
          }

```
And I have tried and it works for situation `fail -> retry-pass -> retry-fail -> retry-pass -> retry-fail`.
Having nested if is much easier to follow.

Reply:
Fixed it